soroban-env-host/src/auth.rs

```diff
@@ -0,0 +1,940 @@
+        let sub_invocations_xdr = xdr_invocation.sub_invocations.to_vec();
+        let sub_invocations = sub_invocations_xdr
+            .into_iter()
+            .map(|i| AuthorizedInvocation::from_xdr(i))
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(Self {
+            contract_id: xdr_invocation.contract_id.clone(),
+            function_name: Symbol::try_from(xdr_invocation.function_name)?,
+            args: xdr_invocation.args.clone(),
+            sub_invocations,
+            is_exhausted: false,
+        })
+    }
+
+    fn to_xdr(&self, budget: &Budget) -> Result<xdr::AuthorizedInvocation, HostError> {
+        Ok(xdr::AuthorizedInvocation {
+            contract_id: self.contract_id.metered_clone(budget)?,
+            // This ideally should be infallible
+            function_name: self
+                .function_name
+                .to_string()
+                .try_into()
+                .map_err(|_| HostError::from(ScUnknownErrorCode::General))?,
+            args: self.args.metered_clone(budget)?,
+            sub_invocations: self
+                .sub_invocations
+                .iter()
+                .map(|i| i.to_xdr(budget))
+                .collect::<Result<Vec<xdr::AuthorizedInvocation>, HostError>>()?
+                .try_into()
+                .map_err(|_| HostError::from(ScUnknownErrorCode::General))?,
+        })
+    }
+
+    fn new_recording(contract_id: &Hash, function_name: Symbol, args: ScVec) -> Self {
+        Self {
+            contract_id: contract_id.clone(),
+            function_name,
+            args,
+            sub_invocations: vec![],
+            is_exhausted: true,
+        }
+    }
+
+    // Non-metered conversion should only be used for the recording preflight
+    // runs.
+    fn to_xdr_non_metered(&self) -> Result<xdr::AuthorizedInvocation, HostError> {
+        Ok(xdr::AuthorizedInvocation {
+            contract_id: self.contract_id.clone(),
+            // This ideally should be infallible
+            function_name: self
+                .function_name
+                .to_string()
+                .try_into()
+                .map_err(|_| HostError::from(ScUnknownErrorCode::General))?,
+            args: self.args.clone(),
+            sub_invocations: self
+                .sub_invocations
+                .iter()
+                .map(|i| i.to_xdr_non_metered())
+                .collect::<Result<Vec<xdr::AuthorizedInvocation>, HostError>>()?
+                .try_into()
+                .map_err(|_| HostError::from(ScUnknownErrorCode::General))?,
+        })
+    }
+
+    // Walks a path in the tree defined by `invocation_id_in_call_stack` and
+    // returns the last visited authorized node.
+    fn last_authorized_invocation_mut(
+        &mut self,
+        invocation_id_in_call_stack: &Vec<Option<usize>>,
+        call_stack_id: usize,
+    ) -> Option<&mut AuthorizedInvocation> {
```

Comment:
Why is this an `Option`?

Reply:
The task of the tracker is to map the 'condensed' tree of authorizations to the current call stack. This function returns the bottom-most matching authorization tree node. Since the current call stack may not have any matches for this particular tracker there won't be any node to return here, hence this is an option.

FWIW in the recording mode we lazily create trackers on first invocation (thus making it the tree root), so this should never return none there. But for the enforcing mode it's perfectly valid to return `None` here as the trackers are created based on the transaction field.

Reply:
But this function only ever returns `Some(self)` right? I think you're referring to `AuthorizationTracker::last_authorized_invocation_mut` and not `AuthorizedInvocation::last_authorized_invocation_mut` (where I made this comment).

Reply:
Sorry, you're right. That's just a refactoring artifact.